content/articles/api-authentication-with-django-knox-and-postman-testing/index.md

```diff
@@ -0,0 +1,343 @@
+    message = models.CharField(max_length=500)
+    owner = models.ForeignKey(User, related_name="leads", null=True, on_delete=CASCADE)
+
+    def __str__(self):
+        return self.name
+```
+Now let us migrate the model into the database. First of all,  we will be generating a migration folder and finally migrate to the default `Sqlite3`.
+
+```bash
+$ python manage.py makemigrations
+$ python manage.py migrate
+```
+
+We have to serialize the data from the leads model.Serialization of data means turning the model data into the `JSON` format. So here is the serialization code. 
+
+```python
+from rest_framework import serializers
+from leadsapi.models import LeadModel
+
+class LeadSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = LeadModel
+        fields = '__all__'
+```
+
+The fields having `__all__` means that all the properties of the lead shall be serialized.
+
+Inside the same `leadsapi` app, make an `api.py`. Here we will be inheriting from the `viewsets` of the DRF and map the serializer class. Also, we will be defining the `permission_class` from the rest framework and get the query set of all leads with the method called `get_querysets`. The code below does the implementation.
+
+```python
+from rest_framework import serializers, viewsets, permissions
+
+from leadsapi.models import LeadModel
+from .serializers import LeadSerializer
+
+class LeadViewset(viewsets.ModelViewSet):
+    serializer_class = LeadSerializer
+    permission_classes = [
+        permissions.IsAuthenticated
+    ]
+
+    def get_queryset(self):
+        return self.request.user.leads.all()
+    
+    def perform_create(self, serializer):
+        serializer.save(owner=self.request.user)
+```
+
+All things have been good so far. The very next thing is to make use of the `urls.py` file inside the `leadsapi` app. 
+
+In this file, we will be making use of the rest framework routers by registering the endpoints.  
+
+
+```python
+from rest_framework import routers
+
+from .api import LeadViewset
+
+router = routers.DefaultRouter()
+router.register('leads', LeadViewset, 'leads')
+
+urlpatterns = router.urls
+```
+Whenever there are `HTTP GET` calls with `/leads` endpoint, the API fetches the client all serialized leads.
+
+Let us register the endpoint with the project by going into the `urls.py` of the `Leads`folder.
+
+```python
+from django.contrib import admin
+from django.urls import path, include
+
+urlpatterns = [
+    path('api/', include('leadsapi.urls')),
+    path('', include('accounts.urls'))
+]
+```
+Note that the `account.urls` shall be used later in the tutorial.
+
+### Securing the endpoints with Django-Knox
+
+Now that we have built the rest API, it is time to secure it by allowing only the authenticated users to perform HTTP `GET`, `POST` methods.
+
+We will make a fresh app called `account` which I believe we have demonstrated above. Follow the procedures, make the new app and add it to the settings.
```

Comment:
Change to " We will make a new app called `account` which I believe we have demonstrated above. Follow the procedures, make the new app and add it to the settings."